experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp

```diff
@@ -0,0 +1,661 @@
+
+      // A LSQ with size 1 or 0 does not make sense and cant be generated by the
+      // LSQ Generator, a minimum size of 2 is needed
+      maxLoadSize = std::max(maxLoadSize, (unsigned)2);
+      maxStoreSize = std::max(maxStoreSize, (unsigned)2);
+
+      handshake::LSQDepthAttr lsqDepthAttr = handshake::LSQDepthAttr::get(
+          mod.getContext(), maxLoadSize, maxStoreSize);
+      setDialectAttr(lsqOp, lsqDepthAttr);
+    }
+  }
+}
+
+std::optional<LSQSizingResult> HandshakeSizeLSQsPass::sizeLSQsForCFDFC(
+    handshake::FuncOp funcOp, llvm::SetVector<unsigned> cfdfcBBs,
+    TimingDatabase timingDB, unsigned initialII, std::string collisions) {
+
+  AdjListGraph graph(funcOp, cfdfcBBs, timingDB, initialII);
+  std::vector<mlir::Operation *> loadOps =
+      graph.getOperationsWithOpName("handshake.lsq_load");
+  std::vector<mlir::Operation *> storeOps =
+      graph.getOperationsWithOpName("handshake.lsq_store");
+
+  if (loadOps.size() == 0 && storeOps.size() == 0)
+    return std::nullopt;
+
+  // Find starting node, which will be the reference to the rest
+  std::tuple<mlir::Operation *, StartTimes> startNodeAndTimes =
+      findStartTimes(graph);
+  mlir::Operation *startNode = get<0>(startNodeAndTimes);
+  StartTimes startTimes = get<1>(startNodeAndTimes);
+
+  // Find Phi node of each BB
+  std::unordered_map<unsigned, mlir::Operation *> phiNodes =
+      getPhiNodes(graph, startNode);
+
+  // connect all phi nodes to the lsq ops in their BB
+  insertAllocPrecedesMemoryAccessEdges(graph, loadOps, phiNodes);
+
+  // connect all start node candidates, with a latency of their shifting time,
+  // for the case that there is no path from the start node to the start node
+  // candidate and that the start time of the candidate is not zero
+  insertStartnodeShiftingEdges(graph, startNode, startTimes);
+
+  // Calculate the earliest start times for each operation in the graph. This is
+  // needed to find the allocation times
+  graph.setEarliestStartTimes(startNode);
+
+  std::vector<unsigned> IIs;
+
+  // The collisions argument can be passed in the compile script
+  // For now there are 3 defined cases: full, half, none
+  // full: determines the worst case II, for the case that all lsq ops have
+  // memory collisions
+  // half: determines the worst case II and alternate it with
+  // the best case II (from buffer placement), for the case that half of the lsq
+  // ops have memory collisions
+  // none: gets the best case II(from buffer
+  // placement), for the case that no lsq ops have memory collisions
+  //
+  // It would also easily be possible to add more cases, by just pushing the IIs
+  // into the IIs vector in the order they are occuring
+
+  if (collisions == "full") {
+    IIs.push_back(graph.getWorstCaseII());
+  } else if (collisions == "half") {
+    IIs.push_back(initialII);
+    IIs.push_back(graph.getWorstCaseII());
+  } else {
+    // none is the default case
+    IIs.push_back(initialII);
+  }
```

Comment:
I don't get how ASAP or SDC-based scheduling relates to this. The best-case II is passed from buffer-placement, I don't know how buffer placement does it. But in my experience, this II was not consistent with the one observed in the simulation (even without any memory dependencies). However, I also don't know if the one from buffer placement is the optimal one. I can just add the note though.

Reply:
> The best-case II is passed from buffer-placement,
> But in my experience, this II was not consistent with the one observed in the simulation (even without any memory dependencies).

Ok good to know. Maybe also leave a note here on the inconsistency between the simulation II vs. the II obtained in MILP. 

What if you calculate the best case II the way you did for the worst case? How does it compare with the simulation II?

Reply:
I added a comment in the code regarding the inconsistency. 

I did not try that. The calculation for the worst-case II is straightforward, but I am not sure how that would work for the best-case II. Unfortunately, I currently also don't have the time to look further into this. If it would be very important I could potentially check it out in late January/February.